packages/lisk-transactions/src/transactions/base.ts

```diff
@@ -369,44 +413,95 @@ export abstract class BaseTransaction {
 		const updatedAccount = { ...sender, balance: updatedBalance.toString() };
+		const errors = updatedBalance.gte(0)
+			? []
+			: [
+					new TransactionError(
+						`Account does not have enough LSK: ${sender.address}, balance: ${
+							// tslint:disable-next-line no-magic-numbers
+							new BigNum(sender.balance.toString() || '0').div(Math.pow(10, 8))
+						}`,
+						this.id,
+					),
+			  ];
 
 		return {
 			id: this.id,
-			status: Status.OK,
-			state: [updatedAccount],
-			errors: [],
+			status: errors.length > 0 ? Status.FAIL : Status.OK,
+			state: { sender: updatedAccount },
+			errors,
 		};
 	}
 
-	public undo(sender: Account): TransactionResponse {
+	public undo({ sender }: RequiredState): TransactionResponse {
 		const updatedBalance = new BigNum(sender.balance).add(this.fee);
 		const updatedAccount = { ...sender, balance: updatedBalance.toString() };
+		const errors = updatedBalance.lte(MAX_TRANSACTION_AMOUNT)
+			? []
+			: [new TransactionError('Invalid balance amount', this.id)];
 
 		return {
 			id: this.id,
-			status: Status.OK,
-			state: [updatedAccount],
-			errors: [],
+			status: errors.length > 0 ? Status.FAIL : Status.OK,
+			state: { sender: updatedAccount },
+			errors,
 		};
 	}
 
 	public isExpired(date: Date = new Date()): boolean {
 		// tslint:disable-next-line no-magic-numbers
 		const timeNow = Math.floor(date.getTime() / 1000);
-		const timeOut = this.isMultisignature
-			? UNCONFIRMED_MULTISIG_TRANSACTION_TIMEOUT
-			: UNCONFIRMED_TRANSACTION_TIMEOUT;
+		const timeOut =
+			this.isMultisignature === MultisignatureStatus.TRUE
+				? UNCONFIRMED_MULTISIG_TRANSACTION_TIMEOUT
+				: UNCONFIRMED_TRANSACTION_TIMEOUT;
 		const timeElapsed =
 			// tslint:disable-next-line no-magic-numbers
 			timeNow - Math.floor(this.receivedAt.getTime() / 1000);
 
 		return timeElapsed > timeOut;
 	}
+
+	protected getBasicBytes(): Buffer {
+		const transactionType = Buffer.alloc(BYTESIZES.TYPE, this.type);
+		const transactionTimestamp = Buffer.alloc(BYTESIZES.TIMESTAMP);
+		transactionTimestamp.writeIntLE(this.timestamp, 0, BYTESIZES.TIMESTAMP);
+
+		const transactionSenderPublicKey = hexToBuffer(this.senderPublicKey);
+
+		const transactionRecipientID = this.recipientId
+			? bigNumberToBuffer(this.recipientId.slice(0, -1), BYTESIZES.RECIPIENT_ID)
+			: Buffer.alloc(BYTESIZES.RECIPIENT_ID);
+
+		const transactionAmount = this.amount.toBuffer({
+			endian: 'little',
+			size: BYTESIZES.AMOUNT,
+		});
+
+		return Buffer.concat([
+			transactionType,
+			transactionTimestamp,
+			transactionSenderPublicKey,
+			transactionRecipientID,
+			transactionAmount,
+			this.getAssetBytes(),
+		]);
+	}
+
+	public sign(passphrase: string, secondPassphrase?: string): void {
+		this._signature = undefined;
+		this._signSignature = undefined;
+		this._signature = signData(hash(this.getBasicBytes()), passphrase);
+
+		if (secondPassphrase) {
+			this._signSignature = signData(
+				hash(this.getBasicBytes()),
```

Comment:
need to include signature bytes?

Reply:
yes! it's fixed in this PR https://github.com/LiskHQ/lisk-elements/pull/998